File: src/main/GO/leetcode/lists/offer/Offer44.go
package offer

import "strconv"

/**
 * @author wujingxinit@outlook.com
 * @date 2023/1/4 23:27
 */

func findNthDigit(n int) int {
	if n == 0 {
		return 0
	}
	start, base, digitCount := 1, 9, 1

	for n > base*digitCount {
		n -= base * digitCount
		base *= 10
		start *= 10
		digitCount++
	}
	start += (n - 1) / digitCount
	num := strconv.Itoa(start)

	return int(num[(n-1)%digitCount] - '0')
}

File: src/main/java/leetcode/lists/offer/Offer44.java
package leetcode.lists.offer;

/**
 * @author jingxinwu
 * @date 2021-11-22 10:58 下午
 */
public class Offer44 {

    public static void main(String[] args) {
        //1 2 3 4 5 6 7 8 9 1 0 1 1 1 2 1 3
        System.out.println(new Offer44().findNthDigit(29));
    }

    public int findNthDigit(int n) {
        int start = 1;
        int digitNum = 1;
        long base = 9;
        while (n > digitNum * base) {
            n -= digitNum * base;
            start *= 10;
            base *= 10;
            digitNum++;
        }
        start += (n - 1) / digitNum;
        String s = String.valueOf(start);
        char c = s.charAt((n - 1) % digitNum);
        return c - '0';
    }
}